### PR TITLE
Unify parsing part of BigInteger with CoreLib

### PR DESCRIPTION
**Pull request by huoyaoyuan**
Part of #28657. This is my third attempt of working on this. I'd like to keep formatting and further cleanup to follow-up PRs to help reviewing.

This PR does not refactor the algorithm part. It just adapts corelib-style patterns for BigInteger. Reviewing commit by commit is recommended.

### CODE REVIEW

**Comment by huoyaoyuan** [comment]
Also asking a question about formatting code here: the following pattern is heavily used in formatting code to share between UTF8 and UTF16:
```csharp
private static void FormatNumber<TChar>(ref ValueListBuilder<TChar> vlb, ref NumberBuffer number, int nMaxDigits, NumberFormatInfo info) where TChar : unmanaged, IUtfChar<TChar>
```

What's the best approach to share those code out of CoreLib? Using #ifdef?

**Comment by danmoseley** [comment]
Merge conflicts - and then this is reviewable? Looks like the other one went in.

**Comment by huoyaoyuan** [comment]
There's also potential massive conflict with #85392/#86875, and minor dependency with other numeric PRs. Can someone determine an order to review these?

**Comment by huoyaoyuan** [comment]
@tannergooding wanna to discuss how to deal with this.

#86875 uses `IUtf8Char` in parsing, brings the same problem from formatting. `IUtf8Char` can't be used in System.Runtime.Numerics thus the majority of code need to update.

The approach I tried looks like this:
```csharp
#if !SYSTEM_PRIVATE_CORELIB
using TChar = System.Char;
#pragma warning disable SA1121 // Use built-in type alias
#endif

#if SYSTEM_PRIVATE_CORELIB
        internal static unsafe TChar* UInt32ToDecChars<TChar>(TChar* bufferEnd, uint value, int digits) where TChar : unmanaged, IUtfChar<TChar>
#else
        internal static unsafe char* UInt32ToDecChars(char* bufferEnd, uint value, int digits)
#endif


#if SYSTEM_PRIVATE_CORELIB
        private static ReadOnlySpan<TChar> NegativeSign<TChar>(NumberFormatInfo info)
            where TChar : unmanaged, IUtfChar<TChar>
            => info.NegativeSignTChar<TChar>();
#else
        private static ReadOnlySpan<char> NegativeSign<TChar>(NumberFormatInfo info) => info.NegativeSign;
#endif
```

The workaround just works, but isn't expandable if we want to support UTF8 for BigInteger. What do you think about the best approach? Should I ask Stephen or someone else?

File: src/libraries/System.Runtime.Numerics/src/System/Number.BigInteger.cs [hunk 1]
```diff
@@ -342,46 +320,59 @@ internal static bool TryValidateParseStyleInteger(NumberStyles style, [NotNullWh
...
         {
-            if (!TryValidateParseStyleInteger(style, out ArgumentException? e))
+
+            scoped Span<byte> buffer;
+            byte[]? arrayFromPool = null;
+
+            if (value.Length < 255)
             {
-                throw e; // TryParse still throws ArgumentException on invalid NumberStyles
+                buffer = stackalloc byte[value.Length + 1 + 1];
```
Comment (adamsitnik) [inline src/libraries/System.Runtime.Numerics/src/System/Number.BigInteger.cs:346]: IIRC it's recommended to use constant-size stack allocations for small buffers.

@jkotas please correct me if I am wrong

File: src/libraries/System.Runtime.Numerics/src/System/Number.BigInteger.cs [hunk 2]
```diff
@@ -342,46 +320,59 @@ internal static bool TryValidateParseStyleInteger(NumberStyles style, [NotNullWh
...
             }
-
-            BigNumberBuffer bigNumber = BigNumberBuffer.Create();
-            if (!FormatProvider.TryStringToBigInteger(value, style, info, bigNumber.digits, out bigNumber.precision, out bigNumber.scale, out bigNumber.sign))
+            else
             {
-                result = default;
-                return ParsingStatus.Failed;
+                buffer = arrayFromPool = ArrayPool<byte>.Shared.Rent(value.Length + 1 + 1);
```
Comment (adamsitnik) [inline src/libraries/System.Runtime.Numerics/src/System/Number.BigInteger.cs:350]: it's the second time 1 + 1 is being used and as a person who reads the code for the first time it's not obvious to me. Could you please add a comment? Or introduce a const with-self-describing-name and use it in both places?

File: src/libraries/System.Runtime.Numerics/src/System/Number.BigInteger.cs [hunk 3]
```diff
@@ -503,6 +512,10 @@ private static ParsingStatus HexNumberToBigInteger(ref BigNumberBuffer number, o
                     ArrayPool<uint>.Shared.Return(arrayFromPool);
                 }
             }
+
+        FailExit:
```
Comment (adamsitnik) [inline src/libraries/System.Runtime.Numerics/src/System/Number.BigInteger.cs:516]: is using `goto` actually beneficial for the performance here?

File: src/libraries/System.Runtime.Numerics/src/System/Number.BigInteger.cs [hunk 4]
```diff
@@ -640,13 +650,13 @@ bool ProcessChunk(ReadOnlySpan<char> chunkDigits, ref Span<uint> currentBuffer)
...
 
-            ParsingStatus DivideAndConquer(ref BigNumberBuffer number, out BigInteger result)
+            ParsingStatus DivideAndConquer(ref NumberBuffer number, out BigInteger result)
             {
                 Span<uint> currentBuffer;
                 int[]? arrayFromPoolForMultiplier = null;
                 try
                 {
-                    totalDigitCount = Math.Min(number.digits.Length - 1, numberScale);
+                    totalDigitCount = Math.Min(number.DigitsCount, numberScale);
```
Comment (adamsitnik) [inline src/libraries/System.Runtime.Numerics/src/System/Number.BigInteger.cs:659]: Is `DigitsCount` equal to `number.digits.Length - 1` here?

**Review by adamsitnik — CHANGES_REQUESTED** [review]
Please excuse me for our Team not providing any review for so long. We have been hit by a *wonderful* talent redeployment quite hard, and this caused the delay.

Big thanks for removing the code duplication!

I've added some comments and I made it clear which can be ignored (or addressed in separate PR).

Overall the PR looks good, it's very thoughtful. However there are merge conflict so I am going to hit "request changes".

@huoyaoyuan  thank you for your contribution!

**Comment by huoyaoyuan** [comment]
ACK for this PR:

Conflicts are coming from new features/refactors after creation of this PR. I'll redo some moving work to ensure no changes are lost.

Currently I will focus on my another PR first.

**Comment by huoyaoyuan** [comment]
Updated. It really helps that BigInteger hasn't adopt UTF-8, otherwise this will be totally invalidated.

File: src/libraries/Common/src/System/Number.Parsing.Common.cs [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+                   {
+                       '\u2012' or         // Figure Dash
+                       '\u207B' or         // Superscript Minus
+                       '\u208B' or         // Subscript Minus
+                       '\u2212' or         // Minus Sign
+                       '\u2796' or         // Heavy Minus Sign
+                       '\uFE63' or         // Small Hyphen-Minus
+                       '\uFF0D' => true,   // Fullwidth Hyphen-Minus
+                       _ => false
+                   };
```
[the remark this answers is not in the document]
Comment (huoyaoyuan) [reply src/libraries/Common/src/System/Number.Parsing.Common.cs:374]: Let's confirm with BigInteger parsing benchmark.

File: src/libraries/System.Runtime.Numerics/src/System/Number.BigInteger.cs (hunk 4 above)
Comment (huoyaoyuan) [reply src/libraries/System.Runtime.Numerics/src/System/Number.BigInteger.cs:791]: Tests are passing so I assume nothing is wrong. `digits` has a trailing zero in the comment.

**Comment by huoyaoyuan** [comment]
BenchmarkDotNet v0.13.11-nightly.20231126.107, Windows 11 (10.0.22631.2715/23H2/2023Update/SunValley3)
13th Gen Intel Core i9-13900K, 1 CPU, 32 logical and 24 physical cores
.NET SDK 8.0.100
  [Host]     : .NET 8.0.0 (8.0.23.53103), X64 RyuJIT AVX2
  Job-UKWEZQ : .NET 9.0.0 (42.42.42.42424), X64 RyuJIT AVX2
  Job-RXOVAY : .NET 9.0.0 (42.42.42.42424), X64 RyuJIT AVX2

PowerPlanMode=00000000-0000-0000-0000-000000000000  Arguments=/p:EnableUnsafeBinaryFormatterSerialization=true  IterationTime=250.0000 ms
MaxIterationCount=20  MinIterationCount=15  WarmupCount=1

| Method | Job        | Toolchain         | numberString                        | Mean      | Error    | StdDev   | Median    | Min       | Max       | Ratio | Gen0   | Allocated | Alloc Ratio |
|------- |----------- |------------------ |------------------------------------ |----------:|---------:|---------:|----------:|----------:|----------:|------:|-------:|----------:|------------:|
| Parse  | Job-UKWEZQ | \PR\corerun.exe   | -2147483648                         |  39.18 ns | 0.604 ns | 0.565 ns |  39.00 ns |  38.49 ns |  40.28 ns |  0.85 | 0.0016 |      32 B |        0.24 |
| Parse  | Job-RXOVAY | \main\corerun.exe | -2147483648                         |  45.89 ns | 0.420 ns | 0.372 ns |  45.80 ns |  45.40 ns |  46.75 ns |  1.00 | 0.0072 |     136 B |        1.00 |
|        |            |                   |                                     |           |          |          |           |           |           |       |        |           |             |
| Parse  | Job-UKWEZQ | \PR\corerun.exe   | 123                                 |  20.52 ns | 0.149 ns | 0.139 ns |  20.49 ns |  20.31 ns |  20.77 ns |  0.72 |      - |         - |        0.00 |
| Parse  | Job-RXOVAY | \main\corerun.exe | 123                                 |  28.36 ns | 0.384 ns | 0.340 ns |  28.41 ns |  27.57 ns |  28.92 ns |  1.00 | 0.0055 |     104 B |        1.00 |
|        |            |                   |                                     |           |          |          |           |           |           |       |        |           |             |
| Parse  | Job-UKWEZQ | \PR\corerun.exe   | 123456789012(...)901234567890 [200] | 612.07 ns | 4.446 ns | 4.159 ns | 612.18 ns | 605.97 ns | 619.63 ns |  0.73 | 0.0049 |     112 B |        0.11 |
| Parse  | Job-RXOVAY | \main\corerun.exe | 123456789012(...)901234567890 [200] | 833.89 ns | 9.352 ns | 8.748 ns | 830.38 ns | 825.52 ns | 853.02 ns |  1.00 | 0.0504 |     984 B |        1.00 |

File: src/libraries/System.Runtime.Numerics/src/System/Number.BigInteger.cs (hunk 3 above)
Comment (huoyaoyuan) [reply src/libraries/System.Runtime.Numerics/src/System/Number.BigInteger.cs:521]: I don't actually remember where I take this pattern from.
The hex parsing itself can be rewritten with HexConverter in a follow-up PR. I will switch to simplest code.

File: src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems [hunk 6]
```diff
@@ -1334,6 +1333,12 @@
     <Compile Include="$(CommonPath)System\NotImplemented.cs">
       <Link>Common\System\NotImplemented.cs</Link>
     </Compile>
+    <Compile Include="$(CommonPath)System\Number.NumberBuffer.cs">
+      <Link>System\Number.NumberBuffer.cs</Link>
+    </Compile>
+    <Compile Include="$(CommonPath)System\Number.Parsing.Common.cs">
+      <Link>System\Number.Parsing.Common.cs</Link>
```
[the remark this answers is not in the document]
Comment (adamsitnik) [reply src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems:1426]: I am fine with that and I hate to block PRs with nit comments, so 👍

**Review by adamsitnik — APPROVED** [review]
LGTM, thank you for your contribution and patience @huoyaoyuan !

PS. I am jealous of your hardware ;)

**Comment by huoyaoyuan** [comment]
Well earlier today I noticed the UTF-8/16 unification of CoreLib is not shared. This reveals the pain I faced for formatting. Will open a follow-up PR for that.

**Comment by huoyaoyuan** [comment]
Opened #95402 as follow up.

**Pull request merged**